text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+        config.path = "/etc/passwd"
+end;
+```
+
+As it may be very verbose to define an artefact with a complicated script, we created the  `with` statement as a syntax sugar that creates a tremor-script block with a multi-let statement, which in turn returns a record value with all the defined variables.
+
+It is also possible to add `args` which can be provided upon artefact creation/instantiation. 
+
+```
+define http connector my_http
+args
+    required_argument,
+    optional_arg_with_default = "default value"
+with
+    codec = "json",
+    interceptors = [ 
+        {
+            "type": "split",
+            "config": {
+                "split_by": "\n"
+            }
+        },
+        "base64"
+    ],
+    config.host = "localhost",
+    config.port = args.required_argument,
+    err_required = args.optional_arg_with_default
+end;
+```
+
+
+#### Required Tremor Script Changes
+
+In order to make defining config entries in tremor-script convenient, we need to add a few features to tremor-script to make this work reasonably well in a configuration context, where all we want is to return a record value without too much fuss and ceremony:
+
+* Add multi-let statements, that combine multiple `let`s inside a single statement whcih defined the variables therein and returns a record value with all those definitions.
+* Add auto-creation of intermediary path segments in let stetements:
+
+  ```
+  let config.nested.value = 1;
+  ```
+
+  In this case if `config.nested` does not exist we would auto-create it as part of this let as an empty record.
+
+### Pipelines
+
+Pipelines are defined using a unique pipeline id, optionally some arguments and a mandatory query block that embeds a trickle query:
+
+EBNF Grammar:
+
+```
+pipeline      = "define" "pipeline" pipeline_id
+                [ "args" argument_list ]
+                "query" 
+                trickle_query
+                "end"
+pipeline_id   = id
+argument_list = ( id | id "=" expr ) [ "," argument_list ]
+```
+
+Whitespace and newlines are not significant here.
+
+**Example:**
+
+`args` can be referenced within the trickle query via the `args` root. As these are filled upon creation/instantiation
+a pipeline definition becomes a template:
+
+```
+define pipeline my_pipeline
+args
+    required_argument,
+    optional_arg_with_default = "default value"
+query
+    use std::datetime;
+    define tumbling window fifteen_secs
+    with
+        interval = datetime::with_seconds(args.required_argument),
+    end;
+
+    select { "count": aggr::stats::count(event) } from in[fifteen_secs] into out having event.count > 0;
+end; # we probably need to find a different way to terminate an embedded trickle query or find some trick
+```
+
+It might be interesting to be able to load a trickle query from a trickle file.
+To that end we add new config directives to trickle that can define arguments and their default values.
+
+```trickle
+#!config args my_arg = "foo", required_arg
```

Comment:
since directives are usually singular, would it make more sense to have:

```
#!config arg my_arg = "foo"
#!config arg required_arg

Reply:
yeah, looks much nicer.